### ISSUE TITLE
ENH: linalg.eigh_tridiagonal: divide and conquer option

### ISSUE THREAD

**Issue by rfeinman** [scipy/scipy#13982]
When requesting all eigenvalues/eigenvectors of a symmetric tridiagonal matrix, the `eigh_tridiagonal` function uses lapack_driver="stemr" by default. However, the [suggested routine](https://software.intel.com/content/www/us/en/develop/documentation/onemkl-developer-reference-fortran/top/lapack-routines/lapack-least-squares-and-eigenvalue-problem-routines/lapack-least-squares-and-eigenvalue-problem-computational-routines/symmetric-eigenvalue-problems-lapack-computational-routines.html) for the *all* problem is STEVD, a divide and conquer method. In my benchmarks, STEMR is not very efficient for full eigendecomposition of sym tridiagonal matrices, especially as n gets large. Neither are the other driver options in `eigh_tridiagonal`.

I'd like to see an implementation of the STEVD routine, which is not currently offered in scipy.linalg.lapack (although STEV is).

As an alternative, for the case of select="a" you could have `eigh_tridiagonal` call `eig_banded` internally, which uses a divide-and-conquer method for the more general banded matrix problem. In my benchmarks, `eig_banded` is much faster than `eigh_tridiagonal` (with any driver) for full decomposition.

___
For reference, here is a table of runtime results using various techniques from scipy.linalg (eigh_tridiagonal, eig_banded, eigh, eig). At each `n` I've created a symmetric tri-diagonal matrix and then factorized it with 4 different techniques, in order of decreasing specialization. It seems odd that going from tridiagonal -> banded decreases complexity siginificantly, since it is less specialized.

```
[-------------------------- eigendecomposition -------------------------]
                |  symmetric tri-d  |   banded  |  symmetric  |  general 
6 threads: --------------------------------------------------------------
      n =  128  |       1173.3      |    464.0  |     1443.5  |    4349.4
      n =  256  |       4755.9      |   1327.5  |     6611.7  |   19106.3
      n =  512  |      21427.9      |   6683.2  |    33384.7  |   52084.4
      n = 1024  |      83946.0      |  44359.5  |   224708.1  |  409845.6

Times are in microseconds (us).

```

**Comment by ilayn** [comment]
We probably need to add the bandwidth check to `eigh` for this functionality to handle it in a central place instead of generating more functions in the public API.

**Comment by j-bowhay** [comment]
> We probably need to add the bandwidth check to `eigh` for this functionality to handle it in a central place instead of generating more functions in the public API.

I'm slightly confused, I agree that would be nice but all the functions mentioned here already exist. The issue is just asking for a different driver?

**Comment by ilayn** [comment]
Yes not to block anything about this request. What I mean is, just like in the `solve` situation, `eigh_tridiagonal` is quite a niche thing that probably we should not have to start with. But it is there. However, we could still centralize all that functionality in one function like adding `assume_a` or whatever and slowly move users towards there.
